providers/implementations/signature/rsa.c

```diff
@@ -244,33 +324,37 @@ static int rsa_setup_mgf1_md(PROV_RSA_CTX *ctx, const char *mdname,
 {
     size_t len;
     EVP_MD *md = NULL;
+    int mdnid;
 
     if (mdprops == NULL)
         mdprops = ctx->propq;
 
-    if (ctx->mgf1_mdname[0] != '\0')
-        EVP_MD_free(ctx->mgf1_md);
-
     if ((md = EVP_MD_fetch(ctx->libctx, mdname, mdprops)) == NULL) {
         ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_DIGEST,
                        "%s could not be fetched", mdname);
         return 0;
     }
     /* The default for mgf1 is SHA1 - so allow SHA1 */
-    if (digest_rsa_sign_get_md_nid(md, 1) == NID_undef) {
-        ERR_raise_data(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED,
-                       "digest=%s", mdname);
+    if ((mdnid = digest_rsa_sign_get_md_nid(md, 1)) == NID_undef
+        || !rsa_check_padding(ctx, NULL, mdname, mdnid)) {
+        if (mdnid == NID_undef)
+            ERR_raise_data(ERR_LIB_PROV, PROV_R_DIGEST_NOT_ALLOWED,
+                           "digest=%s", mdname);
         EVP_MD_free(md);
         return 0;
     }
-    ctx->mgf1_md = md;
     len = OPENSSL_strlcpy(ctx->mgf1_mdname, mdname, sizeof(ctx->mgf1_mdname));
```

Comment:
This assumes the buffers are the same size.. (Guess that is ok since they are currently).